test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/BusinessMetricsUserAgentTest.java

```diff
@@ -0,0 +1,164 @@
+import software.amazon.awssdk.core.useragent.BusinessMetrics;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonAsyncClient;
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonAsyncClientBuilder;
+import software.amazon.awssdk.services.protocolrestjson.model.PaginatedOperationWithResultKeyResponse;
+import software.amazon.awssdk.services.protocolrestjson.paginators.PaginatedOperationWithResultKeyPublisher;
+import software.amazon.awssdk.services.restjsonwithwaiters.RestJsonWithWaitersAsyncClient;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesRequest;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesResponse;
+import software.amazon.awssdk.services.restjsonwithwaiters.waiters.RestJsonWithWaitersAsyncWaiter;
+
+class BusinessMetricsUserAgentTest {
+    private CapturingInterceptor interceptor;
+
+    private static final String USER_AGENT_HEADER_NAME = "User-Agent";
+    private static final StaticCredentialsProvider CREDENTIALS_PROVIDER =
+        StaticCredentialsProvider.create(AwsBasicCredentials.create("akid", "skid"));
+
+    @BeforeEach
+    public void setup() {
+        this.interceptor = new CapturingInterceptor();
+    }
+
+    @AfterEach
+    public void cleanup() {
+
+    }
+
+    private static Stream<Arguments> inputValues() {
+        return Stream.of(
+            Arguments.of("Default values contain at least a retry metric", Arrays.asList("D"))
+        );
+    }
+
+    @ParameterizedTest(name = "{index} - {0}")
+    @MethodSource("inputValues")
+    void validate_metricsString_forDifferentConfigValues(String description, List<String> expectedMetrics) {
+        ProtocolRestJsonAsyncClientBuilder clientBuilder = asyncClientBuilder();
+
+        assertThatThrownBy(() -> clientBuilder.build().allTypes().join()).hasMessageContaining("stop");
+
+        String userAgent = assertAndGetUserAgentString();
+        expectedMetrics.forEach(expectedMetric -> assertThat(userAgent).matches(METRIC_SEARCH_PATTERN.apply(expectedMetric)));
+    }
+
+    @Test
+    void when_waiterIsUsed_correctMetricIsAdded() {
+        RestJsonWithWaitersAsyncClient asyncClient =
+            RestJsonWithWaitersAsyncClient.builder().region(Region.US_WEST_2).credentialsProvider(CREDENTIALS_PROVIDER)
+                                          .overrideConfiguration(c -> c.addExecutionInterceptor(interceptor)).build();
+        RestJsonWithWaitersAsyncWaiter asyncWaiter = RestJsonWithWaitersAsyncWaiter.builder().client(asyncClient).build();
+
+        try {
+            CompletableFuture<WaiterResponse<AllTypesResponse>> responseFuture =
+                asyncWaiter.waitUntilAllTypesSuccess(AllTypesRequest.builder().integerMember(1).build());
+            assertThat(responseFuture.get().attemptsExecuted()).isEqualTo(1);
+        } catch (Exception e) {
```

Comment:
nit, can we use `assertThatThrownBy()`, like the other tests? It's less error prone in cases where an exception isn't thrown.

Reply:
Fixed